Fix/thicknesses_

AngRodrigues opened the pull request:
## Summary of major changes:

- Thickness outputs from all thickness calculators are now merged
- LPF version pinned to 1.4
- Added folder with datasets in map2loop/_datasets/geodata_files/hamersley. The idea behind this is to have a data loading function, e.g., at the moment `load_hamersley_geology()` returns a gdf with the geology; structure and dtm are available, and others will be added later on. This is quite useful for tests, but should also be useful for when we implement #74. The only thing about this is the location, which implies a long import: `from map2loop._datasets.geodata_files.hamersley import load_hamersley_geology`. Happy to change the file location - does anyone has a suggestion?
- `Minimum fault length` is now a parameter of the config file, and it actually removes the faults with length under defined value. 
- Removed the `project.set_minimum_fault_length` method to avoid users setting the parameter twice. 
- added `all_basal_contacts` object which are abnormal+basal contacts , and `basal_contacts` holds only the basal. Reconstruction of `sampled_basal_contacts` is based on `basal_contacts`
- changed the `thickness_calculator_alpha` to be coherent with the logic of the other thickness calculators
- added tests for some of the functions in `mapdata.py` 
- added tests for `thickness_compute` in `project.py`
- rewrote tests for all thickness calculators independently (now without having to run the project)

## summary of other minor modifications:
- `tmpfile` is now used to create the localGdalfile
- changed to f"" strings where possible
- change from `os.path.join` to `pathlib.Path` where possible
- added geopandas to csv converter in mapdata (it was on a #TODO)
- added check and error message to make sure that all units in `user_defined_stratigraphic_column` are in the geology file (otherwise project creation fails; at least now user has a warning). 
- renamed `distance` to `stratigraphic_distance` in `basal_contacts` creation
- removed `set_` and `get_thickness_calculator`
- rectified minor typos in the thickness calculators 
- added a Linalgerror catch/pass for the InterpolatedStructure 

Fixes #86 
Fixes #110 
Fixes #111 
Fixes #12 

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] Test improvement

## How Has This Been Tested?

The branch is fix/thicknesses_AR_2. 

## Checklist:

- [x] This branch is up-to-date with master
- [x] All gh-action checks are passing
- [x] I have performed a self-review of my own code
- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] My tests run with pytest from the map2loop folder
- [x] New and existing tests pass locally with my changes

AngRodrigues committed:
fix: squashed commits of: - fixing the non-use of minimum_fault_length, - use tmpfile to create temporary files, - added random test for mapdata functions:get/set working projection, - added functionality and comments to basal_contacts_build

AngRodrigues committed:
tests: add more functionality

AngRodrigues committed:
fix: leave as is for now

AngRodrigues committed:
fix: finalise the thickness calculator outputs

AngRodrigues committed:
fix: update tests/ small fixes

AngRodrigues committed:
fix: make sure dependencies are right

AngRodrigues committed:
Merge branch 'master' of https://github.com/Loop3D/map2loop into fix/thicknesses_AR_2

AngRodrigues committed:
fix: update thickness tests//fix minor typos

AngRodrigues committed:
fix: correct some typos/update tests

AngRodrigues committed:
fix: typo in tests

AngRodrigues committed:
Update README.md

AngRodrigues committed:
fix: add the basal_contacts_abnormal_check

lachlangrose commented on map2loop/deformation_history.py:
```diff
@@ -30,7 +30,8 @@ def __init__(self):
         """
-        self.minimum_fault_length_to_export = 500.0
+        
+        self.minimum_fault_length_to_export = 0.0
```
should this be a parameter that can be initialised?

rabii-chaarani replied:
in case it's not user-defined, I think it would be better to automatically calculate the minimum fault length given the scale of the map

lachlangrose commented on map2loop/map2model_wrapper.py:
```diff
@@ -133,9 +133,9 @@ def run(self, verbose_level: VerboseLevel = None):
-            os.path.join(self.map_data.tmp_path, "map2model_data"),
-            os.path.join(self.map_data.tmp_path, "map2model_data", "geology_wkt.csv"),
-            os.path.join(self.map_data.tmp_path, "map2model_data", "faults_wkt.csv"),
+            os.path.join(self.map_data.map2model_tmp_path),
```
Can you used pathlib.Path(self.map_data.map2model_tmp_path)/"geology_wkt.csv"... could also change map_data.map2model_tmp_path to a pathlib object?

AngRodrigues replied:
Map2model needs str as arguments, and doesn't accept Path objects. 

Added for map_data in https://github.com/Loop3D/map2loop/pull/112/commits/8892241829be9c59230d2d213c94dafa93ecdbfd

lachlangrose commented on map2loop/mapdata.py:
```diff
@@ -917,6 +906,20 @@ def parse_fault_map(self) -> tuple:
         faults = geopandas.GeoDataFrame(self.raw_data[Datatype.FAULT]["geometry"])
+
+        # crop by minimum fault length
+        self.minimum_fault_length = 0.0
```
replace with this self.minimum_fault_length = self.config.fault_config.get('minimum_fault_length',0.) unless printing the error message is important. I would suggest don't print in the middle of the code for config parameters but do a check of the config params when they are loaded.

AngRodrigues replied:
Yep, makes sense

lachlangrose commented on map2loop/project.py:
```diff
@@ -190,35 +199,38 @@ def __init__(
         self.map_data.load_all_map_data()
 
         # If flag to save out data is check do so
+        tmp_path = pathlib.Path(tempfile.mkdtemp())
```
If the user wants to save the data wouldn't they want to specify where? I agree with using a tempfile directory when the data is just tmp and not saved. But if the user is actually wanting to keep the data they need to be able to save the data. My other question is, if map2loop is run multiple times do you want the temp directory to stay the same or should it be a new temp directory per run? I actually think the temp directory made by tempfile should be create on module import not on create project to prevent creating too many temp files.

AngRodrigues replied:
Yep, good catch; there should not be a tempfile in proj. Changed with https://github.com/Loop3D/map2loop/pull/112/commits/6b954663e6601503d80bcce4564a2fe535254a36
The user can override this path with proj(tmp_path="MyPath")

lachlangrose commented on map2loop/project.py:
```diff
@@ -190,35 +199,38 @@ def __init__(
-
+        
+        # only define the minimum fault length if there are faults
+        if len(self.map_data.get_map_data(Datatype.FAULT))!=0:
```
Isn't the fault minimul length stored in the config?

AngRodrigues replied:
yes, this was redundant. removed in https://github.com/Loop3D/map2loop/pull/112/commits/23873fa16c5d9a7d41fdfff22d1ecc9461c31fb4

lachlangrose commented on map2loop/project.py:
```diff
@@ -351,27 +344,15 @@ def get_sampler(self, datatype: Datatype):
         """
-        return self.deformation_history.get_minimum_fault_length()
-
+        return self.map_data.minimum_fault_length
```
remove? or return the config minimum fault length?

AngRodrigues replied:
removed. now a method in map_data (makes more sense)

lachlangrose commented on map2loop/project.py:
```diff
@@ -446,18 +430,54 @@ def calculate_stratigraphic_order(self, take_best=False):
+            
+        thickness_combined_results = pandas.concat(thickness_calculation_results, axis=1)
+
+        self.stratigraphic_column.stratigraphicUnits.loc[:, 'ThicknessMean_ThicknessCalculatorAlpha'] = thickness_combined_results['ThicknessMean_ThicknessCalculatorAlpha']
```
Could you not just put the thickness_combined_results directly inside the stratigraphicUnits inside the for loop. This means you aren't hard coding the number of thickness calculators

lachlangrose commented on map2loop/stratigraphic_column.py:
```diff
@@ -40,9 +40,15 @@ def __init__(self):
+                ("ThicknessStdDev_InterpolatedStructure", float),
+                ("ThicknessMean_StructuralPoint", float),
+                ("ThicknessMedian_StructuralPoint", float),
+                ("ThicknessStdDev_StructuralPoint", float),
```
Is there a way to make this generic so that the user can specify thickness calculators to keep map2loop being modular?

lachlangrose commented on map2loop/thickness_calculator.py:
```diff
@@ -261,9 +260,18 @@ def compute(
+        
+        ### TODO: RC - fix the Linalg error at some stage. 
+        # for now, if error is raised, return thicknesses with -1
+        try:
```
Is the linalg error coming from colocated points?

AngRodrigues replied:
the collocated points error has been fixed in the meantime. Removing this code.

lachlangrose reviewed (changes requested):
I haven't reviewed or tested the code but I have made some suggestions within the code.

My main comment for this PR is that the  changes to the thickness calculators have not been implemented in the modular framework of map2loop. Instead of setting the list of thickness calculators within the method calling the thickness calculators, the list of thickness calculators should be managed by the project class. The default list can be hard coded using the current three, but the user should be able to call set_thickness_calculators and provide a custom list of thickness calculators. 

This would mean, if the thicknesss calculators have any parameters e.g. the threshold angle for the structure point, or the interpolator type for the interpolated structure these could be set by the user and the thickness calculator objects can be set.

rabii-chaarani committed:
fix: added check for collocated points

AngRodrigues committed:
fix: issue 122

AngRodrigues committed:
Merge branch 'master' of https://github.com/Loop3D/map2loop into fix/thicknesses_AR_2

AngRodrigues committed:
fix: upd gdal as per master

AngRodrigues committed:
fix: remove test from server

AngRodrigues committed:
fix: update str format

rabii-chaarani committed:
fix: now use DBSCAN to aggregate collocated poins

rabii-chaarani committed:
fix: removed addition

rabii-chaarani committed:
fix: added warning when collocated points detected

AngRodrigues committed:
Merge branch 'master' of https://github.com/Loop3D/map2loop into fix/thicknesses_AR_2

AngRodrigues committed:
fix: add str for paths - gdal does not accept pathlibs

AngRodrigues committed:
Merge branch 'master' of https://github.com/Loop3D/map2loop into fix/thicknesses_AR_2

AngRodrigues committed:
fix: update minimum_fault_length as per code review

AngRodrigues committed:
fix: add pathlibs where possible in m2model_wrapper

AngRodrigues committed:
fix: add pathlib in mapdata.py

AngRodrigues committed:
fix: merge coll points branch and remove linalg catch

AngRodrigues committed:
fix: revert pathlib in map2model.run

AngRodrigues committed:
fix: proj should not have temp files

AngRodrigues committed:
fix: update the minimum_fault_length

AngRodrigues committed:
tests: add tests for minimum_fault_length

AngRodrigues committed:
fix: remove redundancy from project.py

AngRodrigues committed:
fix: update calculate minimum_fault_length for 5% of the map area

AngRodrigues committed:
fix: linting

AngRodrigues committed:
fix: typo

AngRodrigues committed:
fix: remove repetitive print stat

AngRodrigues committed:
fix: adjust thickness_calc workflow for modularity

AngRodrigues committed:
fix: remove unnecessary fields

AngRodrigues committed:
fix: update thickness tests

AngRodrigues committed:
Merge branch 'master' of https://github.com/Loop3D/map2loop into fix/thicknesses_AR_2

AngRodrigues committed:
fix: add ignore_fault_codes and tests for ignore_lithology_code and ignore_fault_code

rabii-chaarani committed:
fix: try relative path ../_datasets

rabii-chaarani committed:
fix: use only path instead of pkg_resources

rabii-chaarani committed:
fix: try full path

rabii-chaarani committed:
fix: load only using path /hamersley

rabii-chaarani committed:
fix: try ./map2loop/_datasets/geodata_files/

rabii-chaarani committed:
fix: use correct path

rabii-chaarani committed:
refactor: back to original code

AngRodrigues committed:
fix: add the whole hamersley folder

AngRodrigues committed:
fix: update docker for docs

AngRodrigues commented:
Documentation workflows fail because this branch depends on changes in LPF as well (which is still on a branch; not released yet - https://github.com/Loop3D/LoopProjectFile/tree/fix_thicknesses).

lachlangrose commented:
I think the documentation doesn't need to be built for all branches. We can
make it only run on master release. I am also not sure whether
documentation build/deploy should just be a manually run item that is
triggered on release but can also just be run manually. This could be the
same issue as the package building, it really only needs to run on master
not on other branches.

On Thu, Oct 31, 2024 at 1:43 PM AngRodrigues ***@***.***>
wrote:

> Documentation workflows fail because this branch depends on changes in LPF
> as well (which is still on a branch; not released yet -
> https://github.com/Loop3D/LoopProjectFile/tree/fix_thicknesses).
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Loop3D/map2loop/pull/112#issuecomment-2448908486>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABYHZADESHR4NJ5YLOXPQOLZ6GKODAVCNFSM6AAAAABKRYVZQWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDINBYHEYDQNBYGY>
> .
> You are receiving this because your review was requested.Message ID:
> ***@***.***>
>

AngRodrigues committed:
fix: update the docs-deploy for master branch only

AngRodrigues committed:
fix: update CI to build on master only

AngRodrigues commented:
> I think the documentation doesn't need to be built for all branches. We can make it only run on master release. I am also not sure whether documentation build/deploy should just be a manually run item that is triggered on release but can also just be run manually. This could be the same issue as the package building, it really only needs to run on master not on other branches.

@lachlangrose, I agree. Added these changes to the workflow in this PR; left the sdist jobs so that the tests run on the branch.

AngRodrigues committed:
fix: remove redundancy

AngRodrigues committed:
fix: update the server test and handle timeout properly

AngRodrigues committed:
fix: updated ReadMe to have the right install steps (temporary measure)

AngRodrigues committed:
fix: update CI - test wheels on master only

AngRodrigues committed:
fix: add test to sdist build in CI to ensure tests are run in the non-master branches

AngRodrigues committed:
fix: update pytest in ci

AngRodrigues committed:
fix: revert ci

AngRodrigues committed:
fix: add pytest in the right place

AngRodrigues committed:
fix: update pytest location

AngRodrigues committed:
fix: manifest update

AngRodrigues committed:
fix: update ci again

AngRodrigues committed:
investigate

AngRodrigues committed:
find the install folder

AngRodrigues committed:
try again

AngRodrigues committed:
update ci

AngRodrigues committed:
fix: add checkout step to build sdist

AngRodrigues committed:
fix: run pytest with test sdist action

AngRodrigues committed:
fix: revert CI and let tests only run on main branch

AngRodrigues committed:
fix: add version check new class to check install of  dependencies.txt; includes version number check

AngRodrigues committed:
fix: organise manifest

AngRodrigues commented:
# Edits on 1/11/2024:

- updated methods for ```minimum_fault_length``` 
- minimum fault length calculated from project's bbox, if not defined by user in the ```config_dictionary```. 
- thickness outputs can be still merged, but modularity has been added as suggested. User can set up to 5 thickness calculators (default is ```InterpolatedStructure```). 
- Methods ```set_thickness_calculator``` and ```get_thickness_calculator``` are available again, and used likeso: 

```
from map2loop.thickness_calculator import ThicknessCalculatorAlpha, InterpolatedStructure, StructuralPoint

[..set up the project..]

project.set_thickness_calculator([StructuralPoint(), InterpolatedStructure()])
project.get_thickness_calculator()

Output:
['StructuralPoint', 'InterpolatedStructure']
```

- LPF stores results and name of the thickness calculator as a new field in the LPF. This works with the LPF branch here - https://github.com/Loop3D/LoopProjectFile/tree/fix_thicknesses
- Added ignore codes both for lithology and faults (works by feature name, and similar to the existing method). Added methods for ```set``` and ```get``` with ```_ignore_lithology_codes()``` and  ```_ignore_fault_codes()```. Example of working example code to achieve this can be seen in ```map2loop/tests/project/test_ignore_codes_setters_getters.py```. 
- Added and updated appropriate tests for all new added features. 
- updated CI to run docs and builds workflows only on master branch.  
- added small fixes for:
* Fixes #143
* Fixes #137 
* Fixes #119 

@lachlangrose @rabii-chaarani the PR is ready for re-review now. 
Thank you!
